### ISSUE TITLE
RFC: Ad hoc polymorphism.

### ISSUE THREAD

**Issue by mbudiu-vmw** [vmware/differential-datalog#670]

**Comment by ryzhyk** [comment]
We already have support for argument polymorphism, e.g.:

```
function vec_length(v: Vec<'X>): usize
```

It will be nice to also support ad hoc polymorphism, so that we can define functions with the same name for different types, e.g.,:

```
function length(v: Vec<'X>): usize {...}
function length(v: Set<'X>): usize {...}
function length(v: Map<X>): usize {...}
...
```

I propose adding a limited form of the latter, using the first argument of a function as disambiguator, e.g.,:

```
// (1)
function length(v: Vec<'X>): usize {...}

// ERROR: let's not allow this even though the number of arguments could be used as disambiguator.
function length(v: Vec<'X>, foo: 'Y): usize {...}

// OK
function length(v: Set<'X>): usize {...}

// ERROR: conflicts with (1) 
function length(v: Vec<bool>): usize {...}
...
```

This should not be too hard to implement and will improve DDlog code:
- We will no longer need to prefix function names with type or module names
- Once we support object-oriented function call notation, it will be nicer to write `x.length()` vs `x.vec_length()`

**Comment by ryzhyk** [comment]
@mbudiu-vmw , @blp

**Comment by blp** [comment]
I think I'm in favor of this. It would let the callers look nicer, and I don't see any big pitfalls.

It's sad that `Vec<bool>` couldn't be supported as a specialization of `Vec<'X>`, but I guess that opens a big can of worms.

Of course Mihai knows a lot more about languages....

**Comment by ryzhyk** [comment]
Notes on how this was implemented:

- In the implementation, I ended up allowing the number of arguments to be used as disambiguator, so it is ok to have functions with the same name and first argument as long as they take different numbers of arguments.

- Another decision that had to be made is when to detect conflicting declarations.  Within a module, we always report conflicts regardless of whether conflicting functions are invoked anywhere in the program.  When conflicting declarations appear in different modules, we postpone the check until the user makes a function call that cannot be unambiguously resolved to one of the candidates, i.e., we won't complain even if the user imports both modules unqualified unless they try calling the function.

- Rust does not allow functions with the same name unless they appear inside different modules or `impl` blocks. Both options don't work well for us. Instead the compiler mangles names of polymorphic functions by adding the type of the first argument and the number of arguments to the name. This is invisible to the user, except in `extern` functions, which must be named-mangled manually. This is ugly and fragile; hence the preferred solution is to make sure that extern function names are unique, and define aliases to them in DDlog.  E.g, in `std.dl` we keep old function names like `string_substr` and add new functions, e.g., `substr(string,...)` that invoke them.
